File: validBST.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self.valid(root, None, None)
    
    def valid(self, root: Optional[TreeNode], left, right) -> bool:
        if not root:
            return True
        if left and root.val <= left:
            return False
        if right and root.val >= right:
            return False
        
        return self.valid(root.left, left, root.val) and self.valid(root.right, root.val, right)

File: validBST.js
function TreeNode(val, left, right) {
    this.val = (val===undefined ? 0 : val)
    this.left = (left===undefined ? null : left)
    this.right = (right===undefined ? null : right)
}

var isValidBST = function(root) {
    return valid(root, null, null);

    function valid(node, left, right) {
        if (node == null) return true;
        if (left != null && node.val >= left) return false;
        if (right != null && node.val <= right) return false;

        return valid(node.left, left, node.val) && valid(node.right, node.val, right);
    }
};